File: backend/organization-service/src/main/java/com/pangeranvalerensco/orchestria/organization_service/service/DivisionTaskService.java
package com.pangeranvalerensco.orchestria.organization_service.service;

import java.util.List;

import com.pangeranvalerensco.orchestria.organization_service.entity.enums.TaskStatus;
import com.pangeranvalerensco.orchestria.organization_service.payload.request.DivisionTaskRequest;
import com.pangeranvalerensco.orchestria.organization_service.payload.response.ApiResponse;
import com.pangeranvalerensco.orchestria.organization_service.payload.response.DivisionTaskResponse;

public interface DivisionTaskService {

    ApiResponse<List<DivisionTaskResponse>> getAllTasks();

    ApiResponse<List<DivisionTaskResponse>> getTasksByDivision(Long divisionId);

    ApiResponse<List<DivisionTaskResponse>> getTasksByAssignedMember(Long memberId);

    ApiResponse<List<DivisionTaskResponse>> getTasksByStatus(TaskStatus status);

    ApiResponse<DivisionTaskResponse> getTaskById(Long id);

    ApiResponse<DivisionTaskResponse> createTask(DivisionTaskRequest request);

    ApiResponse<DivisionTaskResponse> updateTask(Long id, DivisionTaskRequest request);

    ApiResponse<DivisionTaskResponse> updateTaskStatus(Long id, TaskStatus status);

    ApiResponse<Void> deleteTask(Long id);

    ApiResponse<List<DivisionTaskResponse>> getMyTasks();

    ApiResponse<DivisionTaskResponse> updateMyTaskStatus(Long id, TaskStatus status);
}

File: frontend/src/services/divisionTaskService.ts
import { apiRequest } from '../api/http';
import type { DivisionTask, DivisionTaskRequest, DivisionTaskEvidence, DivisionTaskEvidenceRequest, TaskStatus } from '../types/divisionTask';

const API_URL = '/api/organization/division-tasks';
const EVIDENCE_API_URL = '/api/organization/division-task-evidences';

const divisionTaskService = {
  // --- Tasks (Manager) ---
  getAllTasks: async (token: string) => {
    return await apiRequest<DivisionTask[]>(API_URL, { method: 'GET' }, token);
  },

  getTasksByDivision: async (token: string, divisionId: number) => {
    return await apiRequest<DivisionTask[]>(`${API_URL}/division/${divisionId}`, { method: 'GET' }, token);
  },

  createTask: async (token: string, data: DivisionTaskRequest) => {
    return await apiRequest<DivisionTask>(API_URL, { method: 'POST', body: JSON.stringify(data) }, token);
  },

  updateTask: async (token: string, id: number, data: DivisionTaskRequest) => {
    return await apiRequest<DivisionTask>(`${API_URL}/${id}`, { method: 'PUT', body: JSON.stringify(data) }, token);
  },

  updateTaskStatus: async (token: string, id: number, status: TaskStatus) => {
    return await apiRequest<DivisionTask>(`${API_URL}/${id}/status/${status}`, { method: 'PATCH' }, token);
  },

  deleteTask: async (token: string, id: number) => {
    return await apiRequest<void>(`${API_URL}/${id}`, { method: 'DELETE' }, token);
  },

  getTaskById: async (token: string, id: number) => {
    return await apiRequest<DivisionTask>(`${API_URL}/${id}`, { method: 'GET' }, token);
  },

  // --- Tasks (Member / Me) ---
  getMyTasks: async (token: string) => {
    return await apiRequest<DivisionTask[]>(`${API_URL}/me`, { method: 'GET' }, token);
  },

  updateMyTaskStatus: async (token: string, id: number, status: TaskStatus) => {
    return await apiRequest<DivisionTask>(`${API_URL}/${id}/my-status/${status}`, { method: 'PATCH' }, token);
  },

  // --- Evidences ---
  getEvidencesByTask: async (token: string, taskId: number) => {
    return await apiRequest<DivisionTaskEvidence[]>(`${EVIDENCE_API_URL}/task/${taskId}`, { method: 'GET' }, token);
  },

  createEvidence: async (token: string, data: DivisionTaskEvidenceRequest) => {
    return await apiRequest<DivisionTaskEvidence>(EVIDENCE_API_URL, { method: 'POST', body: JSON.stringify(data) }, token);
  },

  updateEvidence: async (token: string, id: number, data: DivisionTaskEvidenceRequest) => {
    return await apiRequest<DivisionTaskEvidence>(`${EVIDENCE_API_URL}/${id}`, { method: 'PUT', body: JSON.stringify(data) }, token);
  },

  deleteEvidence: async (token: string, id: number) => {
    return await apiRequest<void>(`${EVIDENCE_API_URL}/${id}`, { method: 'DELETE' }, token);
  },

  createMyEvidence: async (token: string, data: DivisionTaskEvidenceRequest) => {
    return await apiRequest<DivisionTaskEvidence>(`${EVIDENCE_API_URL}/my`, { method: 'POST', body: JSON.stringify(data) }, token);
  },

  updateMyEvidence: async (token: string, id: number, data: DivisionTaskEvidenceRequest) => {
    return await apiRequest<DivisionTaskEvidence>(`${EVIDENCE_API_URL}/my/${id}`, { method: 'PUT', body: JSON.stringify(data) }, token);
  },

  deleteMyEvidence: async (token: string, id: number) => {
    return await apiRequest<void>(`${EVIDENCE_API_URL}/my/${id}`, { method: 'DELETE' }, token);
  }
};

export default divisionTaskService;
